### PR TITLE
Additional changes for the core profiler plugins page

### LINKED ISSUE #38436: Core Profiler - additional changes for the plugins page
**Issue Description:**

There is an open [PR](https://github.com/woocommerce/woocommerce/pull/38405) for the plugin page already, but it requires additional changes that depend on other in-progress issues. 

Please work on the following changes:

1. Test the installation process further, especially with large plugins like Jetpack. 
2. Refactor [InstallAndActivatePlugins](https://github.com/woocommerce/woocommerce/pull/38405/files#diff-cb652db0708acedc660f9bbad201e2546aaa856290d466b086200f265c5dfbc8) and extract it to a state
3. Use `install-activate-plugins from https://github.com/woocommerce/woocommerce/pull/38434
4. `free-extensions` feed has been updated to add `label` and `learn_more` link in https://github.com/woocommerce/woocommerce/pull/38418. Use both values.
5. Confirm tracks
6. Redirect to the existing Jetpack connection page on WPCOM
7. Add installation_priority and use it in the frontend.

### PR DESCRIPTION
**Pull request by moon0326**
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes #38436  .

This PR addresses changes listed in the issue, except for the Jetpack redirection and `InstallAndActivatePlugins` refactor. I'll handle Jetpack redirection in a separate issue, as it requires a bit of context switching.  @rjchow is working on refactoring `InstallAndActivatePlugins` in https://github.com/woocommerce/woocommerce/pull/38577. For the reference, my previous attempt at refactoring is https://github.com/woocommerce/woocommerce/commit/94e548cec115ef4c4a9ed2739b76c542882c4d56

### How to test the changes in this Pull Request:

For the label and description, please refer to Figma design Y5pUYSJPsGEud1vknUZhi8-fi-300-14366 or use the screenshot: 
image

Make sure to turn off `Show Suggestions` from WooCommerce -> Settings -> Advanced -> WooCommerce.com to show the local version of the suggestions.

1. Go to the Plugins page
2. Confirm the plugins have the correct label and description.
3. Confirm the plugins have a `learn more link
4. Select some plugins and click `Continue`
5. Confirm the plugins get installed in order of [install_priority](https://github.com/woocommerce/woocommerce/compare/update/38436-additional-changes-for-plugins-page?expand=1#diff-f508723b8f4a18b29a46152942aa4b58d21a5ddf06866e1238a734bd458f4eb6)
6. Confirm the change requests from the label and description

For now, I've set `install_priority` value by plugin size. 

| Plugin  |  Priority | Size|
| ------------- | ------------- | ------------- |
| Tiktok for Business  | 1  |  54 KB |
| Pinterest for WooCommerce  | 2  | 985 KB |
| WooCommerce Shipping| 3 | 1.6 MB |
| WooCommerce Tax | 4| 1.6 MB|
| WooCommerce Payments| 5| 5 MB|
| Google Listing and Ads|6|7.2MB|
|Mailpoet|7|7.4MB|
|Jetpack|8|25.7MB|

### CODE REVIEW

**Review by ilyasfoo — COMMENTED** [review]
I'm not sure what I'm doing wrong, but I wasn't able to get to the plugins page in core profiler. After `business_info` step, I get redirected straight to loading screen 🤔

**Comment by moon0326** [comment]
> I'm not sure what I'm doing wrong, but I wasn't able to get to the plugins page in core profiler. After `business_info` step, I get redirected straight to loading screen 🤔

My apologies 🙏 We need to turn off `Show Suggestions` to use the updated extension list.

I've updated the test instructions.

```
Make sure to turn off `Show Suggestions` from
WooCommerce -> Settings -> Advanced -> WooCommerce.com 
to show the local version of the suggestions.
```

File: plugins/woocommerce-admin/client/core-profiler/style.scss
Comment (ilyasfoo) [inline plugins/woocommerce-admin/client/core-profiler/style.scss]: I'm not sure if this is supposed to be here? `gray-900` is also missing `$` if so.

File: plugins/woocommerce/src/Internal/Admin/RemoteFreeExtensions/DefaultFreeExtensions.php
Comment (ilyasfoo) [inline plugins/woocommerce/src/Internal/Admin/RemoteFreeExtensions/DefaultFreeExtensions.php:830]:
```suggestion
				'description'      => __( "Accept credit cards and other popular payment methods smoothly.", 'woocommerce' ),
```

**Review by ilyasfoo — COMMENTED** [review]
Thanks for working on this, @moon0326! I wasn't able to review the code properly since I think there's some unrelated commits in the PR, perhaps I'll defer reviewing them after RJ's PR is merged.

I'll review the UI and functionality first:

1. It seems the extension is displayed as `Installed` even though it's deactivated. Is this intended?
1. In smaller screens, `Learn more` should be hidden ([screenshot](https://github.com/woocommerce/woocommerce/assets/3747241/d77bb0d0-d6d9-4297-8308-47119cab6bdc))
1. I think it's also good to arrange the extensions as displayed in Figma, cc @verofasulo if you intend the extensions arrangement should follow design, or could it be random?
1. Probably not related to this PR in particular, I observed some plugins were installed but not activated. I selected multiple plugins (Jetpack, MailPoet, GLA, Pinterest, TikTok, Shipping) and only some plugins were activated (Pinterest, TikTok). The action scheduler job `woocommerce_plugins_install_callback` was complete. There was some PHP warnings in `debug.log` related to Pinterest, no fatal errors.

File: plugins/woocommerce/src/Internal/Admin/RemoteFreeExtensions/DefaultFreeExtensions.php
Comment (moon0326) [reply plugins/woocommerce/src/Internal/Admin/RemoteFreeExtensions/DefaultFreeExtensions.php:830]: Thank you for the catch! I should have mentioned it in detail. Veronica suggested a few more changes on Y5pUYSJPsGEud1vknUZhi8-fi-1698-36100 and the copy change was one of them.

File: plugins/woocommerce-admin/client/core-profiler/style.scss
Comment (moon0326) [reply plugins/woocommerce-admin/client/core-profiler/style.scss]: 🙏 Thank you for the catch.

**Comment by moon0326** [comment]
Thank you for the review!

> It seems the extension is displayed as Installed even though it's deactivated. Is this intended?

I think so, but let's check with @verofasulo ✋ 

> In smaller screens, Learn more should be hidden 

Updated in https://github.com/woocommerce/woocommerce/pull/38616/commits/cb255e748bd2ee7226341212a1fa67cfac446b47

> I think it's also good to arrange the extensions as displayed in Figma, cc @verofasulo if you intend the extensions arrangement should follow design, or could it be random?

According to Y5pUYSJPsGEud1vknUZhi8-fi-1698-36088 I think the order is correct.

Do you see different order by any chance?

Screen Shot 2023-06-07 at 12 19 17 PM

> Probably not related to this PR in particular, I observed some plugins were installed but not activated. I selected multiple plugins (Jetpack, MailPoet, GLA, Pinterest, TikTok, Shipping) and only some plugins were activated (Pinterest, TikTok). The action scheduler job woocommerce_plugins_install_callback was complete. There was some PHP warnings in debug.log related to Pinterest, no fatal errors.

~I'll double check on that.~ You're correct on this. I've copied the refactored version of [InstallAndActivatePlugins](https://github.com/woocommerce/woocommerce/pull/38577/files#diff-cb652db0708acedc660f9bbad201e2546aaa856290d466b086200f265c5dfbc8) from RJ's PR, which doesn't use [InstallAndActivatePluginsAsync](https://github.com/woocommerce/woocommerce/pull/38616/files#diff-4ece682b94e36afbfcaea1f08207ca8bef913b2b49e2923b9cf11b3da7192451R469). I'll review RJ's PR today so that he can merge it, allowing me to rebase and update the code in this PR to use installAndActivatePlugins.

File: plugins/woocommerce-admin/client/core-profiler/style.scss
Comment (moon0326) [reply plugins/woocommerce-admin/client/core-profiler/style.scss]: Fixed in https://github.com/woocommerce/woocommerce/pull/38616/commits/afa71876e125e61b0cc9671edf01247e9deb5466

**Comment by verofasulo** [comment]
> if you intend the extensions arrangement should follow design, or could it be random?

@ilyasfoo Yes, please, let's follow the order in the designs 🙂 The marketing team has reviewed it and wants to showcase the extensions in that particular order.

**Comment by verofasulo** [comment]
> It seems the extension is displayed as Installed even though it's deactivated. Is this intended?
I think so, but let's check with @verofasulo ✋

@moon0326 I'm not sure what we're referring to 🙂 Could you please clarify? Thanks!

**Comment by ilyasfoo** [comment]
> @moon0326 I'm not sure what we're referring to 🙂 Could you please clarify? Thanks!

Hey, @verofasulo! Moon was referring to the extensions having `Installed` label when it's installed, but not activated. In my opinion, I think we have to label it `Installed` only when it's both **installed & activated**. If the plugin is inactive, the user may not know how to activate it, thus it could become broken UX down the line.

image

**Review by ilyasfoo — COMMENTED** [review]
> According to Y5pUYSJPsGEud1vknUZhi8-fi-1698-36088 I think the order is correct.
> 
> Do you see different order by any chance?

My bad, I may have referred to an outdated design. Sorry! Thanks for addressing the learn more change. Functionality wise this is looking good! I'll try to review them as soon as RJ's PR is merged and rebased

**Comment by verofasulo** [comment]
> Moon was referring to the extensions having Installed label when it's installed, but not activated. In my opinion, I think we have to label it Installed only when it's both installed & activated. If the plugin is inactive, the user may not know how to activate it, thus it could become broken UX down the line.

@ilyasfoo, good point! I didn't think about this specific scenario but it makes sense. Let's show the label only if the extension is active 🙂 Thank you 🙏

**Comment by moon0326** [comment]
Thanks @verofasulo  and @ilyasfoo ❤️ 

@ilyasfoo  // I've rebased the latest main, which has RJ's changes.

**Review by ilyasfoo — COMMENTED** [review]
I'm getting the same issue with @rjchow [here](https://github.com/woocommerce/woocommerce/pull/38577#issuecomment-1583916549), where plugins list is inconsistent. I think I may have found the way to reproduce it:

1. Go through the profiler until the plugins step.
2. In another tab or CLI, delete the option `woocommerce_default_country` or replace with empty string.
3. Refresh the plugins list.
4. Observe the plugin list is not complete.

I'm also getting this error when pressing `Continue` in the plugins page after the rebase:

```
Uncaught TypeError: Cannot read properties of undefined (reading 'find')
    at installAndActivatePlugins.ts:198:43
    at Array.sort (<anonymous>)
    at pluginsInstallationQueue (installAndActivatePlugins.ts:197:41)
    at utils.js:417:1
    at Array.reduce (<anonymous>)
    at updateContext (utils.js:399:1)
    at handleAction (actions.js:584:41)
    at processBlock (actions.js:625:1)
    at Array.map (<anonymous>)
    at resolveActions (actions.js:650:32)
```

Also from my previous review, the plugin is still labelled as `Installed` when deactivated.

**Comment by moon0326** [comment]
> I'm getting the same issue with @rjchow https://github.com/woocommerce/woocommerce/pull/38577#issuecomment-1583916549, where plugins list is inconsistent. I think I may have found the way to reproduce it:

Thank you for the link. I totally missed the ping from the PR. @rjchow Are you still able to reproduce the issue? I've tried calling the endpoint multiple times and it all came back the same for me.

> In another tab or CLI, delete the option woocommerce_default_country or replace with empty string.

As far as I know, `woocommerce_default_country` always has a value. The plugin list is incomplete because some plugins are only available for certain countries. I think the behavior is expected when `woocommerce_default_country` value is missing.

> I'm also getting this error when pressing Continue in the plugins page after the rebase:

@ilyasfoo Fixed in  https://github.com/woocommerce/woocommerce/pull/38616/commits/fcb6b67d993f4b5fe7d2921524b9023a2338a17a 

> Also from my previous review, the plugin is still labelled as Installed when deactivated.

Fixed in https://github.com/woocommerce/woocommerce/pull/38616/commits/e09db7e89f7c20c7fceab954d4aa6dca72f0143b

**Comment by rjchow** [comment]
> > I'm getting the same issue with @rjchow [#38577 (comment)](https://github.com/woocommerce/woocommerce/pull/38577#issuecomment-1583916549), where plugins list is inconsistent. I think I may have found the way to reproduce it:
> 
> Thank you for the link. I totally missed the ping from the PR. @rjchow Are you still able to reproduce the issue? I've tried calling the endpoint multiple times and it all came back the same for me.
> 
> > In another tab or CLI, delete the option woocommerce_default_country or replace with empty string.
> 
> As far as I know, `woocommerce_default_country` always has a value. The plugin list is incomplete because some plugins are only available for certain countries. I think the behavior is expected when `woocommerce_default_country` value is missing.

If that's the case I suspect what's going on is that the getFreeExtensions() dispatch is happening before the [persistBusinessInfo](https://github.com/woocommerce/woocommerce/blob/97a250b870a9f8fe2f47f9201e7f80f9bde82ef5/plugins/woocommerce-admin/client/core-profiler/index.tsx#L935)
call completes

I can look into it when I'm back from AFK tomorrow

File: packages/js/data/src/onboarding/actions.ts
Comment (ilyasfoo) [inline packages/js/data/src/onboarding/actions.ts:468]: I think it'll be good to document the async behaviour, i.e.: how it works and how it's handled in the frontend. Not sure if this is the best place, but I digress. What do you think?

File: packages/js/data/src/onboarding/types.ts
Comment (ilyasfoo) [inline packages/js/data/src/onboarding/types.ts]:
```suggestion
	status: 'pending' | 'in-progress' | 'completed' | 'failed';
```

File: plugins/woocommerce-admin/client/core-profiler/index.tsx
Comment (ilyasfoo) [inline plugins/woocommerce-admin/client/core-profiler/index.tsx]: It seems `recordTracksPluginsLearnMoreLinkClicked` isn't defined anywhere? 🤔

File: plugins/woocommerce/src/Internal/Admin/RemoteFreeExtensions/DefaultFreeExtensions.php
Comment (ilyasfoo) [inline plugins/woocommerce/src/Internal/Admin/RemoteFreeExtensions/DefaultFreeExtensions.php:825]: Perhaps we should comment about how how `install_priority` is handled here such as the ordering since I think maintainers going to need to know what to assign when adding new extensions.

File: plugins/woocommerce/src/Internal/Admin/RemoteFreeExtensions/DefaultFreeExtensions.php
Comment (ilyasfoo) [inline plugins/woocommerce/src/Internal/Admin/RemoteFreeExtensions/DefaultFreeExtensions.php]:
```suggestion
		// Copy shipping for the core-profiler and remove is_visible conditions, except for the country restriction.
```

Comment (ilyasfoo) [reply plugins/woocommerce/src/Internal/Admin/RemoteFreeExtensions/DefaultFreeExtensions.php]: Hey, @moon0326. Do you have the context for why we need to copy shipping and remove `is_visible`?

Wouldn't it be better if we apply the change in the `get_plugin` function instead?

File: plugins/woocommerce-admin/client/core-profiler/services/installAndActivatePlugins.ts
Comment (ilyasfoo) [inline plugins/woocommerce-admin/client/core-profiler/services/installAndActivatePlugins.ts:201]: I'd love it if there's a comment on the sorting that is done here, i.e: ascending or descending

**Review by ilyasfoo — CHANGES_REQUESTED** [review]
Thanks for fixing them, @moon0326! I confirmed that plugins are considered `Installed` only when it's activated.

> Probably not related to this PR in particular, I observed some plugins were installed but not activated. I selected multiple plugins (Jetpack, MailPoet, GLA, Pinterest, TikTok, Shipping) and only some plugins were activated (Pinterest, TikTok). The action scheduler job woocommerce_plugins_install_callback was complete. There was some PHP warnings in debug.log related to Pinterest, no fatal errors.

I'm still having the above issue, it seems only 2 out of 5 plugins installed get activated. Not sure why, the scheduled action is completed. I don't mind to have this addressed in a follow-up if you feel like it though.

Not sure if this is expected, but it seems we're using the old APIs, `/wp-json/wc-admin/plugins/install?_locale=user` and `/wp-json/wc-admin/plugins/activate?_locale=user`? 🤔 

image

Some minor comments:

File: packages/js/data/src/onboarding/types.ts
Comment (moon0326) [reply packages/js/data/src/onboarding/types.ts]: Thank you for the typo checks 🙏

File: plugins/woocommerce/src/Internal/Admin/RemoteFreeExtensions/DefaultFreeExtensions.php
Comment (moon0326) [reply plugins/woocommerce/src/Internal/Admin/RemoteFreeExtensions/DefaultFreeExtensions.php]: @ilyasfoo I think we should do that in the future when we have completely removed the profile wizard. I would like to preserve the existing code as much as possible in case we encounter an issue and need to switch back to the profile wizard.

File: plugins/woocommerce/src/Internal/Admin/RemoteFreeExtensions/DefaultFreeExtensions.php
Comment (moon0326) [reply plugins/woocommerce/src/Internal/Admin/RemoteFreeExtensions/DefaultFreeExtensions.php:825]: Good idea. I'll copy the table from this PR to the comment.

File: plugins/woocommerce-admin/client/core-profiler/index.tsx
Comment (moon0326) [reply plugins/woocommerce-admin/client/core-profiler/index.tsx]: 🤦 I think I missed it from the last rebase.

File: packages/js/data/src/onboarding/actions.ts
Comment (moon0326) [reply packages/js/data/src/onboarding/actions.ts:468]: I think we can document it in both frontend and backend. I'll add them 👍

**Comment by moon0326** [comment]
> I'm still having the above issue, it seems only 2 out of 5 plugins installed get activated. Not sure why, the scheduled action is completed. I don't mind to have this addressed in a follow-up if you feel like it though.

You're correct. I did not change the endpoint after the last rebase 🤦 ~I'll handle it first thing tomorrow morning.~

Updated in https://github.com/woocommerce/woocommerce/pull/38616/commits/5fb686353da6f8a69179f3383266f3a41a39f615

File: plugins/woocommerce/src/Internal/Admin/RemoteFreeExtensions/DefaultFreeExtensions.php
Comment (moon0326) [reply plugins/woocommerce/src/Internal/Admin/RemoteFreeExtensions/DefaultFreeExtensions.php:824]: Updated in https://github.com/woocommerce/woocommerce/pull/38616/commits/31f8600ae5bc465e5481b836fe9b1ec2bbc8a3fb

File: plugins/woocommerce-admin/client/core-profiler/index.tsx
Comment (moon0326) [reply plugins/woocommerce-admin/client/core-profiler/index.tsx]: Updated in https://github.com/woocommerce/woocommerce/pull/38616/commits/31f8600ae5bc465e5481b836fe9b1ec2bbc8a3fb

**Comment by moon0326** [comment]
@ilyasfoo I've addressed missing recordTracksPluginsLearnMoreLinkClicked and activation issues. 

Could we address the documentation changes in a follow-up PR?

File: plugins/woocommerce-admin/client/core-profiler/actions/tracks.tsx
Comment (ilyasfoo) [inline plugins/woocommerce-admin/client/core-profiler/actions/tracks.tsx]: Do we need to track which extension "Learn more" was clicked on? cc @verofasulo if you have input 🙏

File: plugins/woocommerce/src/Internal/Admin/RemoteFreeExtensions/DefaultFreeExtensions.php
Comment (ilyasfoo) [reply plugins/woocommerce/src/Internal/Admin/RemoteFreeExtensions/DefaultFreeExtensions.php]: Makes sense 👍

**Review by ilyasfoo — APPROVED** [review]
> @ilyasfoo I've addressed missing recordTracksPluginsLearnMoreLinkClicked and activation issues.

Thanks! Tests well.

> Could we address the documentation changes in a follow-up PR?

Sounds good! 🚢

File: plugins/woocommerce-admin/client/core-profiler/actions/tracks.tsx
Comment (moon0326) [reply plugins/woocommerce-admin/client/core-profiler/actions/tracks.tsx]: I think we should. I'll add it 👍

Comment (moon0326) [reply plugins/woocommerce-admin/client/core-profiler/actions/tracks.tsx]: Updated in https://github.com/woocommerce/woocommerce/pull/38616/commits/175deece5529fb4c1c6710ed72cd9f34bad3ef6f

**Pull request merged**
